src/sos_filter.c

```diff
@@ -0,0 +1,158 @@
+    struct sos_filter_section filter[0];
+};
+
+inline uint8_t
+overflows_int32(int64_t value) {
+    return value > (int64_t)INT32_MAX || value < (int64_t)INT32_MIN;
+}
+
+// Multiply a coefficient in fixedQ_coeff_t by a value fixedQ_value_t
+static inline fixedQ_value_t
+fixed_mul(struct sos_filter *sf, const fixedQ_coeff_t coeff
+        , const fixedQ_value_t value) {
+    // This optimizes to single cycle SMULL on Arm Coretex M0+
+    int64_t product = (int64_t)coeff * (int64_t)value;
+    // round up at the last bit to be shifted away
+    product += sf->coeff_rounding;
+    // shift the decimal right to discard the coefficient fractional bits
+    int64_t result = product >> sf->coeff_frac_bits;
+    // check for overflow of int32_t
+    if (overflows_int32(result)) {
+        shutdown("fixed_mul: overflow");
+    }
+    // truncate significant 32 bits
+    return (fixedQ_value_t)result;
+}
+
+// Apply the sosfilt algorithm to a new datapoint
+// returns the fixedQ_value_t filtered value
+int32_t
+sosfilt(struct sos_filter *sf, const int32_t unfiltered_value) {
+    if (!sf->is_active) {
+        shutdown("sos_filter not property initialized");
+    }
+
+    // an empty filter performs no filtering
+    if (sf->n_sections == 0) {
+        return unfiltered_value;
+    }
+
+    fixedQ_value_t cur_val = unfiltered_value;
+    // foreach section
+    for (int section_idx = 0; section_idx < sf->n_sections; section_idx++) {
+        struct sos_filter_section *section = &(sf->filter[section_idx]);
+        // apply the section's filter coefficients to input
+        fixedQ_value_t next_val = fixed_mul(sf, section->coeff[0], cur_val);
+        next_val += section->state[0];
+        section->state[0] = fixed_mul(sf, section->coeff[1], cur_val)
+                            - fixed_mul(sf, section->coeff[3], next_val)
+                            + (section->state[1]);
+        section->state[1] = fixed_mul(sf, section->coeff[2], cur_val)
+                            - fixed_mul(sf, section->coeff[4], next_val);
+        cur_val = next_val;
+    }
+
+    return (int32_t)cur_val;
+}
+
+// Create an sos_filter
+void
+command_config_sos_filter(uint32_t *args)
+{
+    uint32_t max_sections = args[1];
+    uint32_t size = offsetof(struct sos_filter, filter[max_sections]);
+    struct sos_filter *sf = oid_alloc(args[0]
+                            , command_config_sos_filter, size);
+    sf->max_sections = max_sections;
+    sf->is_active = 0;
+}
+DECL_COMMAND(command_config_sos_filter, "config_sos_filter oid=%c"
+    " max_sections=%u");
+
+// Lookup an sos_filter
+struct sos_filter *
+sos_filter_oid_lookup(uint8_t oid)
+{
+    return oid_lookup(oid, command_config_sos_filter);
+}
+
+// Set one section of the filter
+void
+command_sos_filter_set_section(uint32_t *args)
+{
+    struct sos_filter *sf = sos_filter_oid_lookup(args[0]);
+    // setting a section marks the filter as inactive
+    sf->is_active = 0;
+    uint8_t section_idx = args[1];
```

Comment:
For what it is worth, it would be preferable to do more validation on `section_idx` (so as to avoid memory corruption even if the host sends something wacky).

I'll make these changes after merging.